### PR TITLE
swift-module-digester: diagnose adding/removing protocol conformances as API breakages.

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsModuleDiffer.def [hunk 1]
```diff
@@ -58,6 +58,10 @@ ERROR(decl_added,none,"%0 is added to a non-resilient type", (StringRef))
 
 ERROR(default_arg_removed,none,"%0 has removed default argument from %1", (StringRef, StringRef))
 
+ERROR(conformance_removed,none,"%0 has removed conformance to %1", (StringRef, StringRef))
+
+ERROR(conformance_added,none,"%0 has added conformance to %1", (StringRef, StringRef))
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsModuleDiffer.def]: Adding a conformance is not an API or ABI break.

Also I'm assuming this applies to resilient types too, and if the conformance is defined in an extension?

File: tools/swift-api-digester/ModuleAnalyzerNodes.cpp [hunk 2]
```diff
@@ -701,7 +721,16 @@ bool SDKNode::operator==(const SDKNode &Other) const {
       }
       LLVM_FALLTHROUGH;
     }
-    case SDKNodeKind::DeclType:
+    case SDKNodeKind::DeclType: {
```
Comment (slavapestov) [inline tools/swift-api-digester/ModuleAnalyzerNodes.cpp]: What if I'm adding a conformance in an extension, and the extended type is from a different module? You really should model conformances as top-level entities, and not as part of types. Then adding a new conformance is fine (just like adding a new top-level function), but removing one is not allowed.

Comment (nkcsgexi) [reply tools/swift-api-digester/ModuleAnalyzerNodes.cpp]: If the extension is for a type from a different module, the tool will synthesize the type definition in its module dump ([here](https://github.com/apple/swift/blob/master/tools/swift-api-digester/ModuleAnalyzerNodes.cpp#L1189)). So the standalone extension won't be missed.

File: include/swift/AST/DiagnosticsModuleDiffer.def (hunk 1 above)
Comment (nkcsgexi) [reply include/swift/AST/DiagnosticsModuleDiffer.def]: The check performs on all types and if the conformance is defined in extensions. Adding conformances can be a problem for protocols (i.e. adding inherited protools).

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test Linux platform

File: include/swift/AST/DiagnosticsModuleDiffer.def (hunk 1 above)
Comment (slavapestov) [reply include/swift/AST/DiagnosticsModuleDiffer.def]: Please don't model inherited protocols as conformances. Adding an inherited protocol is indeed an API and ABI break, but that's the only case.

A bona-fide "conformance" always involves a concrete type and is its own thing.

Comment (nkcsgexi) [reply include/swift/AST/DiagnosticsModuleDiffer.def]: It's already been scoped to protocol only in TOT and we've reworded the diagnostic message.

**Pull request merged**